File: monolith_backend/steps/petstore/api_step.py
from pydantic import BaseModel
from typing import Optional
from src.services.pet_store import pet_store_service
from src.services.types import Pet

class PetRequest(BaseModel):
    name: str
    photoUrl: str

class FoodOrder(BaseModel):
    id: str
    quantity: int

class RequestBody(BaseModel):
    pet: PetRequest
    foodOrder: Optional[FoodOrder] = None

config = {
    "type": "api",
    "name": "ApiTrigger",
    "description": "basic-tutorial api trigger",
    "flows": ["basic-tutorial"],
    "method": "POST",
    "path": "/basic-tutorial",
    "bodySchema": RequestBody.model_json_schema(),
    "responseSchema": {
        200: Pet.model_json_schema(),
    },
    "emits": ["process-food-order"],
}

async def handler(req, context):
    body = req.get("body", {})
    context.logger.info("Step 01 – Processing API Step", {"body": body})

    pet = body.get("pet", {})
    food_order = body.get("foodOrder", {})
    
    new_pet_record = await pet_store_service.create_pet(pet)

    if food_order:
        await context.emit({
            "topic": "process-food-order",
            "data": {
                "id": food_order.get("id"),
                "quantity": food_order.get("quantity"),
                "email": "test@test.com",  # sample email
                "pet_id": new_pet_record.get("id"),
            },
        })

    return {"status": 200, "body": {**new_pet_record, "traceId": context.trace_id}}

File: services/motia-demo/steps/petstore/api.step.ts
import { ApiRouteConfig, Handlers } from 'motia'
import { z } from 'zod'
import { petStoreService, petSchema } from '../../src/services/pet-store'

export const config: ApiRouteConfig = {
  type: 'api',
  name: 'ApiTrigger',
  description: 'basic-tutorial api trigger',
  flows: ['basic-tutorial'],

  method: 'POST',
  path: '/basic-tutorial',
  bodySchema: z.object({
    pet: z.object({
      name: z.string(),
      photoUrl: z.string(),
    }),
    foodOrder: z
      .object({
        id: z.string(),
        quantity: z.number(),
      })
      .optional(),
  }),
  responseSchema: {
    200: petSchema,
  },
  emits: ['process-food-order'],
}

export const handler: Handlers['ApiTrigger'] = async (req, { logger, traceId, emit }) => {
  logger.info('Step 01 – Processing API Step', { body: req.body })

  const { pet, foodOrder } = req.body
  const newPetRecord = await petStoreService.createPet(pet)

  if (foodOrder) {
    await emit({
      topic: 'process-food-order',
      data: {
        ...foodOrder,
        email: 'test@test.com', // sample email
        petId: newPetRecord.id,
      },
    })
  }

  return { status: 200, body: { ...newPetRecord, traceId } }
}
